File: Lagrange_interpolation.cpp
mint interpolate(std::vector<mint>& dataset, int x) {
    int n = dataset.size();
    std::vector<mint> prods_f(n + 1), prods_b(n + 1);
    prods_f[0] = prods_b[n] = mint::raw(1);
    for(int i = 0; i < n; ++i) {
        prods_f[i + 1] = prods_f[i] * (x - i);
    }
    for(int i = n - 1; i >= 0; --i) {
        prods_b[i] = prods_b[i + 1] * (x - i);
    }
    mint ret = mint::raw(0);
    for(int i = 0; i < n; ++i) {
        int rcnt = n - 1 - i;
        if (rcnt & 1) {
            ret -= prods_f[i] * prods_b[i + 1] * iFact[i] * iFact[rcnt] * dataset[i];
        } else {
            ret += prods_f[i] * prods_b[i + 1] * iFact[i] * iFact[rcnt] * dataset[i];
        }
    }
    return ret;
}

File: Lagrange_interpolation.py
def interpolate(dataset, x):
    n = len(dataset)
    rv = 0
    pf = [1] * (n + 1)
    pb = [1] * (n + 1)
    for i in range(n):
        pf[i+1] = pf[i] * (x - i) % MOD
    for i in range(n - 1, -1, -1):
        pb[i] = pb[i+1] * (x - i) % MOD
    for i in range(n):
        rv += pf[i] * pb[i + 1] * ifact[n - 1 - i] * ifact[i] * (-1 if n - 1 - i & 1 else 1) * dataset[i]
        rv %= MOD
    return rv
